Alh/fetch foo properties

wolfsage opened the pull request:
- Support fetchOtherFoos even if fetchFoos isn't specified
 - Support fetchFooProperties / fetchOtherFooProperties

wolfsage committed:
ix_get_list - Allow fetchOtherFoos when main fetchFoos isn't present

Grab the full objects instead of just ids so that if fetchOtherFoos is
present we have the properties necessary to make those queries.

wolfsage committed:
Add fetchFooProperties / fetchOtherFooProperties

Wherever you can fetchFoos or fetchOtherFoos you can fetchFooProperties or
fetchOtherfooProperties.

wolfsage committed:
ix_finalize - If ix_get_list_enabled is true, enforce the API a bit

wolfsage committed:
Allow rclasses to define their own differ for ix_get_list_updates

This is necessary if they utilize a filter with a cond_builder who's
name does not match a property on the entity type.

rjbs commented on lib/Ix/DBIC/ResultSet.pm:
```diff
@@ -1075,7 +1093,15 @@ sub ix_get_list_updates ($self, $ctx, $arg = {}) {
-        if (differ($entity->$filter, $arg->{filter}{$filter})) {
+        my $diff;
+
+        if (my $differ = $filter_map->{$filter}->{differ}) {
```
I don't see any tests for custom differs.

wolfsage committed:
Stringify val like we used to to prevent custom SQL searches

cond_builders are responsible for ensuring they do something sensible with
the non-stringified version.

wolfsage commented:
Right, need to add some here and to topicbox.

Also, just pushed a5c8f34 which is fairly important (until we validate arguments more strictly)

wolfsage committed:
getFooListUpdates - query on all immutable fields instead of required fields

Some classes may not have any required filters at all (at least, not in
metadata).

To determine which subset of entites we need to look at, we can query
those properties that they passed in that are immutable.

For now, immutable means:

 * not in the list of ix_mutable_properties
 * OR specifies an xref_to

I believe all xref_tos are really immutable, but we don't have a way
to actually make them so in metadata yet. (is_immutable will prevent
rows from being created with those properties passed in).

We really need is_initializable for this.

wolfsage committed:
Add tests for changes in the last commit for getFooListUpdates

 * Make sure our custom differ works
 * Make sure we limit things to immutable fields when we don't have 'required'

rjbs commented:
Thanks.  Yes, I believe this is all improvements.
